### PR TITLE
Preprocess deduplication

### PR DESCRIPTION
**Pull request by denkv**
- [x] resolve merge conflicts

### CODE REVIEW

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/URIGraphSizeBasedScoreCalculator.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        return 1 / (float) uriScore;
+    }
+
+    /**
+     * This method return the number of times the Uri occurs as a subject in the graphs.
+     *
+     * @param uri the Uri for which the score has to calculated
+     * @return number of times the Uri occurs as a subject in the graphs
+     */
+    private int getGraphSize(String uri) {
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/URIGraphSizeBasedScoreCalculator.java:94]: "This method return the number of times the Uri occurs as a subject in the graphs." -> In that case the method (and the class) name does not really fit. Maybe you can come up with a better name, e.g., getSubjectTripleCount or something similar.

File: squirrel.api/src/main/java/org/dice_research/squirrel/queue/UriQueue.java [hunk 2]
```diff
@@ -1,51 +1,59 @@
...
+    public void open();
+
+    /**
+     * Adds the given {@link CrawleableUri} instances to the queue.
+     *
+     * @param uris
+     *            the {@link CrawleableUri} instances that should be added to the
+     *            queue.
+     */
+    public List<CrawleableUri> addUris(List<CrawleableUri> uris);
```
Comment (MichaelRoeder) [inline squirrel.api/src/main/java/org/dice_research/squirrel/queue/UriQueue.java:58]: What does it return? Please extend the javadoc comment.

File: squirrel.frontier/src/test/java/org/dice_research/squirrel/frontier/impl/FrontierImplTest.java [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 import java.net.InetAddress;
 import java.net.URI;
+import java.net.URISyntaxException;
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/test/java/org/dice_research/squirrel/frontier/impl/FrontierImplTest.java:10]: Why is the import added if the class is never used?

File: squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java [hunk 4]
```diff
@@ -162,4 +232,84 @@ public void deleteUris() throws Exception {
         assertEquals(0, mongodbQueue.length());
         mongodbQueue.close();
     }
+
+    @Test
+    public void testMakeURIsKeywise() throws URISyntaxException {
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java:237]: This test method should not be part of this test class since the grouping of URIs is done by a different class than the queue.

File: squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java [hunk 5]
```diff
@@ -162,4 +232,84 @@ public void deleteUris() throws Exception {
...
+        mongodbQueue.addUri(uri2);
+        mongodbQueue.addUri(uri3);
+        dbpediaUris.add(uri1);
+        dbpediaUris.add(uri2);
+        dbpediaUris.add(uri3);
+        keyWiseUris.put("dbpedia.org", dbpediaUris);
+        List<CrawleableUri> urisNotAdded = mongodbQueue.addKeywiseUris(keyWiseUris, 2);
+        Assert.assertEquals(3, urisNotAdded.size());
+        mongodbQueue.purge();
+        mongodbQueue.close();
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java:303]: This test should ask the queue for URIs. If the test case would have this, I am sure that it would reveal that the URIs still have been added because of the bug in the workflow described in one of my previous comments.

File: squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java [hunk 6]
```diff
@@ -162,4 +232,84 @@ public void deleteUris() throws Exception {
...
+
+    @Test
+    public void testURIScore() throws URISyntaxException {
+        mongodbQueue.open();
+        mongodbQueue.purge();
+        float score1 = mongodbQueue.getURIScore(new CrawleableUri(new URI("http://dbpedia.org/resource/Berlin")));
+        float score2 = mongodbQueue.getURIScore(new CrawleableUri(new URI("http://dbpedia.org/resource/Bangalore")));
+        Assert.assertEquals(.1f, score1, .0001);
+        Assert.assertEquals(.125f, score2, .0001);
+    }
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/test/java/org/dice_research/squirrel/queue/domainbased/MongoDBDomainQueueTest.java:314]: The calculation of scores is not a task of the queue and, hence, shouldn't be part of its tests. The test for the class implementing the scoring function should have this test.

File: squirrel.worker/src/main/java/org/dice_research/squirrel/worker/impl/WorkerImpl.java [hunk 7]
```diff
@@ -328,15 +330,17 @@ public void sendNewUris(Iterator<byte[]> uriIterator) {
                 uriProcessor.recognizeUriType(newUri);
                 newUris.add(newUri);
                 if ((newUris.size() >= (packageCount + 1) * MAX_URIS_PER_MESSAGE) && uriIterator.hasNext()) {
+                    List<CrawleableUri> distinctUris = newUris.stream().distinct().collect(Collectors.toList());
```
Comment (MichaelRoeder) [inline squirrel.worker/src/main/java/org/dice_research/squirrel/worker/impl/WorkerImpl.java:333]: This change is not necessary. It is not the task of the worker to ensure that the URIs are distinct. In fact, it will assume that the URI collector did this already. So please undo this change.

File: squirrel.worker/src/main/java/org/dice_research/squirrel/worker/impl/WorkerImpl.java [hunk 8]
```diff
@@ -328,15 +330,17 @@ public void sendNewUris(Iterator<byte[]> uriIterator) {
...
+                            new ArrayList<>(distinctUris.subList(packageCount * MAX_URIS_PER_MESSAGE, newUris.size())));
                     packageCount++;
                 }
             } catch (Exception e) {
                 LOGGER.warn("Couldn't handle the (de-)serialization of a URI. It will be ignored.", e);
             }
         }
-        frontier.addNewUris(newUris);
+        List<CrawleableUri> distinctUris = newUris.stream().distinct().collect(Collectors.toList());
+        frontier.addNewUris(distinctUris);
```
Comment (MichaelRoeder) [inline squirrel.worker/src/main/java/org/dice_research/squirrel/worker/impl/WorkerImpl.java:343]: The same as above: calling the distinct method does not help and will consume a lot of runtime.

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorebasedfilter/IUriKeywiseFilter.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+    /**
+     * This method returns the {@link CrawleableUri}s to be added to the queue after filtering out the ones based on their score
+     *
+     * @param keyWiseUris            map of based on key {@link CrawleableUri}s to be filtered
+     * @param minNumberOfUrisToCheck minimum number of {@link CrawleableUri}s to be checked for their score before filtering
+     * @param criticalScore          the critical score to check for filtering
+     * @return {@link Map} of {@link CrawleableUri}s to be added to the queue with their scores
+     */
+    Map<CrawleableUri, Float> filterUrisKeywise(Map<T, List<CrawleableUri>> keyWiseUris, int minNumberOfUrisToCheck, float criticalScore);
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorebasedfilter/IUriKeywiseFilter.java:21]: I think the signature of the method is not a good choice. It is focused very much on a simple and direct way to implement your score-related features but misses to structure the functionality in a way that enables a future usage of this interface. Apart from that, it creates issues in the methods that use it (as described in one of the queue classes).
* `keyWiseUris`: sure, you want to have this input so that is fine.
* `minNumberOfUrisToCheck`: is this really something that needs to be given as a parameter of this method? You could also define it as an attribute of the filter implementation since not all filter implementations may want to make use of it.
* `criticalScore`: to me, it looks like this is definitely a value that should be an attribute of the filter implementation. It is a classic threshold that is defined once and never changed throughout the different calls of the method.
* Return type `Map<CrawleableUri, Float>`. I think that this choice leads to a lot of issues. The classes that implement this interface are grouping the URIs by a key. But the return type forces them to remove the grouping. Instead, URIs are mapped to a score that 1) may not be possible by other, future implementations of this interface (simply because they may not use a score at all) and 2) could be solved in a much easier way by assigning the score to the single URI objects using the `addData` method.

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorebasedfilter/UriKeywiseFilter.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class filters the {@link CrawleableUri}s to be added to the queue based on the score.
+ */
+public class UriKeywiseFilter implements IUriKeywiseFilter {
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorebasedfilter/UriKeywiseFilter.java:16]: The name is misleading. It is not simply some UriKeywiseFilter. It is mainly based on scores that are calculated for the single URIs. The name should reflect that, e.g., `ScoreBasedUriKeywiseFilter` or something similar. I am sure you can come up with something that fits better :wink:

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/IUriScoreCalculator.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package org.dice_research.squirrel.queue.scorecalculator;
+
+import org.dice_research.squirrel.data.uri.CrawleableUri;
+
+/**
+ *  Interface for getting duplicity score of a Uri.
+ */
+public interface IUriScoreCalculator {
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/IUriScoreCalculator.java:8]: Is it really a duplicity score as described in the javadoc comment? The interface itself simply defines a method to calculate a score for a given URI. So why do you want to reduce its usage to the use case of searching for duplicates?

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/IUriScoreCalculator.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+    /**
+     * Returns the score based on specific score calculation implementation
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/IUriScoreCalculator.java:11]: This comment is already much better since it concentrates on the signature of the method. However, it would be a little bit better if you would include that the score is calculated for the given URI, e.g, "Returns a score for the given URI based on the implemented scoring function."

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/UriScoreCalculator.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+
+/**
+ * Interface for getting duplicity score of a Uri.
+ * The duplicity score is calculated for a {@link CrawleableUri} as 1 / (number of times the Uri occurs as a subject in graphs
+ */
+public class UriScoreCalculator implements IUriScoreCalculator {
```
Comment (MichaelRoeder) [inline squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorecalculator/UriScoreCalculator.java:27]: Here, it is exactly the opposite to my comment in the IUriScoreCalculator interface: this class name pretends that the class calculates a score for a given URI. Although this is true, its functionality is much more specific since it calculates a score that tries to express the chance that the URI is a duplicate of an already known URI. The class name should reflect that.

File: squirrel.api/src/main/java/org/dice_research/squirrel/frontier/Frontier.java [hunk 14]
```diff
@@ -31,17 +31,7 @@
     List<CrawleableUri> getNextUris();
 
     /**
-     * Add this URIs to the {@link Frontier}s internal queue if the internal
-     * rules of the {@link Frontier} allow it.
-     *
-     * @param uri
-     *            the URI that should be added to the {@link Frontier}
-     */
-    void addNewUri(CrawleableUri uri);
```
Comment (MichaelRoeder) [inline squirrel.api/src/main/java/org/dice_research/squirrel/frontier/Frontier.java:40]: The same as for the queue interface: please do not simply remove methods from such a central interface. You can mark it as deprecated and even add a default implementation which uses the addNewUris method.

File: squirrel.frontier/src/main/java/org/dice_research/squirrel/data/uri/norm/NormalizerImpl.java [hunk 15]
```diff
@@ -45,11 +45,11 @@
...
     private final static BitSet UNESCAPED_CHARS = new BitSet(0x7F);
-    private final List<String> sessionIDs;
+    private final Pattern queryArgumentsToRemove;
     private final Map<String, Integer> defaultPortMap;
-    
-    public NormalizerImpl(List<String> sessionIDs, Map<String, Integer> defaultPortMap) {
-    	this.sessionIDs = sessionIDs;
+
+    public NormalizerImpl(List<String> queryArgumentsToRemove, Map<String, Integer> defaultPortMap) {
+    	this.queryArgumentsToRemove = Pattern.compile(String.join("|", queryArgumentsToRemove), Pattern.CASE_INSENSITIVE);
```
[the remark this answers is not in the document]
Comment (denkv) [reply squirrel.frontier/src/main/java/org/dice_research/squirrel/data/uri/norm/NormalizerImpl.java:52]: The pattern is used to match an entire single query argument name, so an empty pattern or an empty list of arguments would just do nothing.
